### PR TITLE
refactor: unified macro for type system

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

**To reviewers: hide whitespaces when reviewing.**

Introduce the unified macro for the type system. All relationships between type terms in our system should be recorded in this macro.

https://github.com/risingwavelabs/risingwave/blob/fd98649808750f3c6638ade56f284bcb40206535/src/common/src/types/macros.rs#L15-L63

Previous `for_all_xx` is now a projection of this macro.

https://github.com/risingwavelabs/risingwave/blob/fd98649808750f3c6638ade56f284bcb40206535/src/common/src/types/macros.rs#L65-L82

Also, introduce several _magical_ (😄) `dispatch` macros. The outstanding part is that it can help us to bind some type aliases to the given identifiers, then we can directly refer to the concrete type of `Array` or `Scalar` in the code block. This should make it much more expressive!

https://github.com/risingwavelabs/risingwave/blob/fd98649808750f3c6638ade56f284bcb40206535/src/common/src/types/macros.rs#L155-L221

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/types/macros.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+/// Dispatch the code block to all variants of `ArrayImpl`.
+///
+/// # Usage
+///
+/// The basic usage to access the inner concrete `impl Array` value is:
+///
+/// ```ignore
+/// fn do_stuff<A: Array>(array: &A) { .. }
+///
+/// fn do_stuff_impl(array_impl: &ArrayImpl) {
```
Comment (stdrc) [inline src/common/src/types/macros.rs:164]: What about `do_stuff_dispatch`? `do_stuff_impl` is somewhat ambiguous, sometime we use `foo_impl` to represent internal implementation of a public function `foo`.

File: src/common/src/types/ordered.rs [hunk 2]
```diff
@@ -17,22 +17,21 @@
...
-        }
-    };
-}
-for_all_scalar_variants!(gen_default_partial_cmp_scalar_ref_impl);
+pub fn default_partial_cmp_scalar_ref_impl(
+    lhs: ScalarRefImpl<'_>,
+    rhs: ScalarRefImpl<'_>,
+) -> Option<Ordering> {
+    dispatch_scalar_ref_variants!(lhs, lhs, [S = ScalarRef], {
+        let rhs: S<'_> = rhs.try_into().ok()?;
```
Comment (stdrc) [inline src/common/src/types/ordered.rs:30]: Will this possibly prevent the `match` from optimization?

File: src/common/src/util/sort_util.rs [hunk 3]
```diff
@@ -448,15 +449,24 @@ pub fn compare_rows_in_chunk(
...
-                match (lhs_array.as_ref(), rhs_array.as_ref()) {
-                    $((ArrayImpl::$variant_name(lhs_inner), ArrayImpl::$variant_name(rhs_inner)) => Ok(compare_values_in_array(lhs_inner, lhs_idx, rhs_inner, rhs_idx, column_order.order_type)),)*
-                    (l_arr, r_arr) => Err(InternalError(format!("Unmatched array types, lhs array is: {}, rhs array is: {}", l_arr.get_ident(), r_arr.get_ident()))),
-                }?
-            }
-        }
-        let res = for_all_variants! { gen_match };
+
+        let res = dispatch_array_variants!(&**lhs_array, lhs_inner, {
+            let rhs_inner = (&**rhs_array).try_into().map_err(|_| {
```
Comment (stdrc) [inline src/common/src/util/sort_util.rs:454]: Ditto.

**Review by stdrc — APPROVED** [review]
Ingenious work! You deserve a title of Master of Decl Macros

File: src/common/src/types/macros.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/types/macros.rs:164]: Okay

File: src/common/src/types/ordered.rs (hunk 2 above)
Comment (BugenZhao) [reply src/common/src/types/ordered.rs:30]: Seems no much difference! 😄

https://godbolt.org/z/fqGvPjKG8

**Pull request merged**